components/script/docs/JS-Servos-only-GC.md

```diff
@@ -0,0 +1,396 @@
+    the lifetime "a as defined on the block at 197:56...</b>
+document.rs:197 fn bogus_get_window<"a>(&amp;self) -> JSRef<"a, Window> {
+document.rs:198     let window = self.window.root();
+document.rs:199     window.root_ref()
+document.rs:200 }
+document.rs:197:57: 200:6 <span style="color: green">note:</span> <b>...but borrowed value is only
+    valid for the block at 197:56</b>
+document.rs:197 fn bogus_get_window<"a>(&amp;self) -> JSRef<"a, Window> {
+document.rs:198     let window = self.window.root();
+document.rs:199     window.root_ref()
+document.rs:200 }</pre>
+
+We also implement the [`Deref` trait][deref] for `JSRef<T>`. This allows us to
+access fields of the underlying type `T` through a `JSRef<T>`. Because `JS<T>`
+does *not* implement `Deref` or otherwise provide access to the underlying
+pointer, we have to root an object before using it.
+
+[deref]: http://doc.rust-lang.org/std/ops/trait.Deref.html
+
+The DOM methods of `Window` (for example) are defined in a trait which is
+[implemented][windowmethods] for `JSRef<Window>`. This ensures that the
+`self` pointer is rooted for the duration of the method call, which would not
+be guaranteed if we implemented the methods on `Window` directly.
+
+XXX: this is not really true, is it?
+
+[windowmethods]: XXX
+
+You can check out the [`js` module's documentation][js-docs] for more details
+that didn't make it into this document.
+
+[js-docs]: http://doc.servo.org/script/dom/bindings/js/index.html
+
+Custom static analysis
+======================
+
+To recap, the safety of our system depends on two major parts:
+
+- The auto-generated `trace` methods ensure that SpiderMonkey's garbage
+  collector can see all of the references between DOM objects.
+- The implementation of `Root<T>` and `JSRef<T>` guarantees that we can't use a
+  DOM object from Rust without telling SpiderMonkey about our temporary
+  reference.
+
+But there's a hole in this scheme. We could copy an unrooted pointer — a
+`JS<T>` — to a local variable on the stack, and then at some later point, root
+it and use the DOM object. In the meantime, SpiderMonkey's garbage collector
+won't know about that `JS<T>` on the stack, so it might free the DOM object.
+To really be safe, we need to make sure that `JS<T>` *only* appears in places
+where it will be traced, such as DOM structs, and never in local variables,
+function arguments, and so forth.
+
+// XXX: this gives me a crazy idea: what if we made JS<T> !Copy?
+
+This rule doesn't correspond to anything that already exists in Rust's type
+system. Fortunately, the Rust compiler can load 'lint plugins' providing custom
+static analysis. These basically take the form of new compiler warnings,
+although in this case we set the default severity to 'error'.
+
+[lints]: https://github.com/rust-lang/rust/pull/15024 (better link?)
+
+We have already [implemented a plugin][js-lint] which simply forbids `JS<T>`
```

Comment:
FWIW the current lint doesn't forbid `JS<T>`, I went a different route.

It forbids:

 - `must_root` types in non-`must_root` structs
 - assignment of local variables to `must_root` types
 - having functions take in `must_root` types